service/external_paysession_test.go

```diff
@@ -1,52 +1,183 @@
+
+		req, err := http.NewRequest("Get", "", nil)
 		So(err, ShouldBeNil)
-		req.Body = ioutil.NopCloser(bytes.NewReader([]byte("invalid_body")))
+
+		q := req.URL.Query()
+		q.Add(":payment_id", "1234")
+		req.URL.RawQuery = q.Encode()
+
 		w := httptest.NewRecorder()
-		CreateExternalPaymentJourney(w, req)
+
+		httpmock.Activate()
+		defer httpmock.DeactivateAndReset()
+		costArray := []models.CostResource{defaultCost}
+		jsonResponse, _ := httpmock.NewJsonResponder(200, costArray)
+		httpmock.RegisterResponder("GET", "http://dummy-resource", jsonResponse)
+
+		mockPaymentService.CreateExternalPaymentJourney(w, req)
 		So(w.Code, ShouldEqual, 400)
 	})
 
-	reqBodyWrongPayment := []byte("{\"payment_method\": \"PayPal\",\"resource\": \"http://dummy-resource\"}")
+	Convey("Error response from GovPay", t, func() {
+		mock := dao.NewMockDAO(mockCtrl)
+		mockPaymentService := createMockPaymentService(mock, cfg)
+
+		mock.EXPECT().GetPaymentResource("1234").Return(&models.PaymentResource{ID: "1234", Data: models.PaymentResourceData{Amount: "10.00", Links: models.Links{Resource: "http://dummy-resource"}, PaymentMethod: "GovPay"}}, nil)
 
-	Convey("Invalid Payment Method", t, func() {
 		req, err := http.NewRequest("Get", "", nil)
 		So(err, ShouldBeNil)
-		req.Body = ioutil.NopCloser(bytes.NewReader(reqBodyWrongPayment))
+
+		q := req.URL.Query()
+		q.Add(":payment_id", "1234")
+		req.URL.RawQuery = q.Encode()
+
 		w := httptest.NewRecorder()
-		CreateExternalPaymentJourney(w, req)
-		So(w.Code, ShouldEqual, 400)
+
+		httpmock.Activate()
+		defer httpmock.DeactivateAndReset()
+		costArray := []models.CostResource{defaultCost}
+		jsonResponse, _ := httpmock.NewJsonResponder(200, costArray)
+		httpmock.RegisterResponder("GET", "http://dummy-resource", jsonResponse)
+
+		httpmock.RegisterResponder("POST", cfg.GovPayURL, httpmock.NewStringResponder(400, "error"))
+
+		mockPaymentService.CreateExternalPaymentJourney(w, req)
+		So(w.Code, ShouldEqual, 500)
+	})
+
+	Convey("No NextURL returned from GovPay", t, func() {
+		mock := dao.NewMockDAO(mockCtrl)
+		mockPaymentService := createMockPaymentService(mock, cfg)
+
+		mock.EXPECT().GetPaymentResource("1234").Return(&models.PaymentResource{ID: "1234", Data: models.PaymentResourceData{Amount: "10.00", Links: models.Links{Resource: "http://dummy-resource"}, PaymentMethod: "GovPay"}}, nil)
+
+		req, err := http.NewRequest("Get", "", nil)
+		So(err, ShouldBeNil)
+
+		q := req.URL.Query()
+		q.Add(":payment_id", "1234")
+		req.URL.RawQuery = q.Encode()
+
+		w := httptest.NewRecorder()
+
+		httpmock.Activate()
+		defer httpmock.DeactivateAndReset()
+		costArray := []models.CostResource{defaultCost}
+		jsonResponse, _ := httpmock.NewJsonResponder(200, costArray)
+		httpmock.RegisterResponder("GET", "http://dummy-resource", jsonResponse)
+
+		IncomingGovPayResponse := models.IncomingGovPayResponse{}
+		govPayJSONResponse, _ := httpmock.NewJsonResponder(201, IncomingGovPayResponse)
```

Comment:
http.statusCreated could be used here as well.

Reply:
Changed both of them

Reply:
https://github.com/companieshouse/payments.api.ch.gov.uk/pull/44/files#diff-1ef3aad51fb2ffb3106ab2df065ea7b2R80

Reply:
Fixed